### PR TITLE
Resize data collections in PROBLEME_A_RESOUDRE

### PR DESCRIPTION
**Pull request by JasonMarechal25**
Some collections in PROBLEME_A_RESOUDRE are optimistically sized early but never resized to their proper size. When exporting those collections to clients by API it unnecessarily.

Resizing for NombreDeVariables could be done earlier but for NombreDeContraintes we need to wait until we have build all constraints with ConstraintBuilder.

ConstraintsMatrixCoeff and ColumnIndexes collections are initially oversized (By a large factor).
Before exporting data through API we properly dimension them and remove extra 0 padding

### CODE REVIEW

File: src/solver/optimisation/opt_construction_matrice_des_contraintes_outils.cpp [hunk 1]
```diff
@@ -30,8 +30,8 @@ void OPT_ChargerLaContrainteDansLaMatriceDesContraintes(
   int NombreDeTermesDeLaContrainte,
   char SensContrainte)
 {
-    int& nombreDeTermes = ProblemeAResoudre->NombreDeTermesDansLaMatriceDesContraintes;
```
Comment (flomnes) [inline src/solver/optimisation/opt_construction_matrice_des_contraintes_outils.cpp]: Does not using a reference here make any difference ?

File: src/solver/optimisation/opt_optimisation_lineaire.cpp [hunk 2]
```diff
@@ -139,6 +139,34 @@ void runThermalHeuristic(PROBLEME_HEBDO* problemeHebdo)
...
+                    unsigned nombreDeContraintes)
+{
+    ProblemeAResoudre->CoutQuadratique.resize(nombreDeVariables);
+    ProblemeAResoudre->CoutLineaire.resize(nombreDeVariables);
+    ProblemeAResoudre->TypeDeVariable.resize(nombreDeVariables);
+    ProblemeAResoudre->Xmin.resize(nombreDeVariables);
+    ProblemeAResoudre->Xmax.resize(nombreDeVariables);
+    ProblemeAResoudre->X.resize(nombreDeVariables);
+    ProblemeAResoudre->AdresseOuPlacerLaValeurDesVariablesOptimisees.resize(nombreDeVariables);
+    ProblemeAResoudre->AdresseOuPlacerLaValeurDesCoutsReduits.resize(nombreDeVariables);
```
Comment (flomnes) [inline src/solver/optimisation/opt_optimisation_lineaire.cpp:175]: nombreDeContraintes, to be checked

File: src/solver/optimisation/opt_optimisation_lineaire.cpp [hunk 3]
```diff
@@ -139,6 +139,34 @@ void runThermalHeuristic(PROBLEME_HEBDO* problemeHebdo)
...
+    ProblemeAResoudre->PositionDeLaVariable.resize(nombreDeVariables);
+    ProblemeAResoudre->Pi.resize(nombreDeVariables);
```
Comment (flomnes) [inline src/solver/optimisation/opt_optimisation_lineaire.cpp]: `nbTermes` for `Pi`

File: src/solver/optimisation/opt_optimisation_lineaire.cpp [hunk 4]
```diff
@@ -139,6 +139,34 @@ void runThermalHeuristic(PROBLEME_HEBDO* problemeHebdo)
...
+    ProblemeAResoudre->Colonne.resize(nombreDeVariables);
```
Comment (flomnes) [inline src/solver/optimisation/opt_optimisation_lineaire.cpp]: `nbTermes` for `Colonne`

File: src/solver/optimisation/opt_construction_matrice_des_contraintes_outils.cpp (hunk 1 above)
Comment (JasonMarechal25) [reply src/solver/optimisation/opt_construction_matrice_des_contraintes_outils.cpp]: Yes. I'm reafecting the result at the end of the scope.
The idea is to have a clear affectation of the variables. With a reference it's kinda hidden

Comment (a-zakir) [reply src/solver/optimisation/opt_construction_matrice_des_contraintes_outils.cpp]: forget it! this file is deleted in develop

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 3 above)
Comment (JasonMarechal25) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: Pi is initially sized with nombreDeVariables in OPT_AllocateFromNumberOfVariableConstraints.
Why resize it to nbThermes ?

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 4 above)
Comment (JasonMarechal25) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: Colonne is initially sized with nombreDeVariables in OPT_AllocateFromNumberOfVariableConstraints.
Why resize it to nbThermes ?

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 3 above)
Comment (flomnes) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: `Pi` and `Colonne` are just temporary variables used to fill the constraint matrix. They should not be used after this step, so there is no need to resize them IMO.

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 4 above)
Comment (flomnes) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: See answer above

**Review by flomnes — APPROVED** [review]

**Pull request merged**